src/Compilers/CSharp/Portable/Compilation/CSharpCompilerDiagnosticAnalyzer.cs

```diff
@@ -27,19 +30,26 @@ internal override CommonMessageProvider MessageProvider
 
         internal override ImmutableArray<int> GetSupportedErrorCodes()
         {
-            var errorCodes = Enum.GetValues(typeof(ErrorCode));
-            var builder = ArrayBuilder<int>.GetInstance(errorCodes.Length);
-            foreach (ErrorCode errorCode in errorCodes)
+            var current = _supportedErrorCodes;
+            if (current.IsDefault)
             {
-                // Compiler diagnostic analyzer does not support build-only diagnostics.
-                if (!ErrorFacts.IsBuildOnlyDiagnostic(errorCode) &&
-                    errorCode is not (ErrorCode.Void or ErrorCode.Unknown))
+                var errorCodes = Enum.GetValues(typeof(ErrorCode));
+                var builder = ArrayBuilder<int>.GetInstance(errorCodes.Length);
+                foreach (ErrorCode errorCode in errorCodes)
                 {
-                    builder.Add((int)errorCode);
+                    // Compiler diagnostic analyzer does not support build-only diagnostics.
+                    if (!ErrorFacts.IsBuildOnlyDiagnostic(errorCode) &&
+                        errorCode is not (ErrorCode.Void or ErrorCode.Unknown))
+                    {
+                        builder.Add((int)errorCode);
+                    }
                 }
+
+                ImmutableInterlocked.InterlockedInitialize(ref _supportedErrorCodes, builder.ToImmutableAndFree());
             }
 
-            return builder.ToImmutableAndFree();
+            Debug.Assert(!_supportedErrorCodes.IsDefault);
+            return _supportedErrorCodes;
```

Comment:
It looks like in the only call site of this method, in `CompilerDiagnosticAnalyzer.SupportedDiagnostics`, we are converting these codes into `DiagnosticDescriptor` and producing a new array. Should the caching be done there instead?

Also, it feels like the method could be made `protected` given this, which might help make it more clear how the method is used.

Reply:
ok.  i do see it.  just that it's a smaller array...  i don't understand, but i agree it makes more sense to fix at the analyzer level.  changing to that.